[BB-933] Allow overriding xblock icons with plugins

Agrendalath opened the pull request:
This is a follow-up to #19772, which allowed setting custom icons for each unit in Studio. By default the icon would be derived from unit's type. The suggestion was to do move it to a plugin. This PR introduces some changes required to move it to external Django application

<s>Most of this diff is about extracting modals from monolithic `cms/static/js/views/modals/course_outline_modals.js` to make them reusable. We also extracted `CourseOutlineView` class into a field in `cms/static/js/views/pages/course_outline.js` to make this view easily extendable.</s>

For overriding implementation of getting icon, we implemented `pluggable_override` decorator.

**JIRA tickets**: [OSPR-3801](https://openedx.atlassian.net/browse/OSPR-3801)

**Merge deadline**: None

**Sandbox**: https://pr21433.sandbox.opencraft.hosting/ (provisioning)

**Testing instructions**:

1. Go to [this page](http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/?child=first) and see that the first unit icon is [fa-book](https://fontawesome.com/v4.7.0/icon/book).
1. Install the `master` version of [edx-django-utils](https://github.com/edx/edx-django-utils) in LMS (`make lms-shell`):
    `pip install -e git+https://github.com/edx/edx-django-utils.git#egg=edx-django-utils`
1. Install [custom-unit-icons](https://github.com/open-craft/custom-unit-icons) for LMS (`make lms-shell`):
    `pip install -e git+https://github.com/open-craft/custom-unit-icons.git@agrendalath/bb-933-address_override_changes#egg=custom-unit-icons`
1. Set the following in `lms/envs/private.py` and run `make lms-restart`:
   ```
   from .common import XBLOCK_MIXINS
   OVERRIDE_GET_UNIT_ICON = 'custom_unit_icons.icons.get_icon'
   XBLOCK_MIXINS += ('custom_unit_icons.icons.IconOverrideMixin',)
   ```
1. Go to [this page](http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/?child=first) and see that the first unit icon is still [fa-book](https://fontawesome.com/v4.7.0/icon/book).
1. Run the following snippet in LMS Django shell:
   ```python
   from xmodule.modulestore.django import modulestore
   from opaque_keys.edx.keys import UsageKey

   usage_key = UsageKey.from_string('block-v1:edX+DemoX+Demo_Course+type@vertical+block@867dddb6f55d410caaa9c1eb9c6743ec')
   item = modulestore().get_item(usage_key)
   item.icon_override = 'video'
   modulestore().update_item(item, 1)
   ```
1. Go to [this page](http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/?child=first) and see that the first unit icon is now [fa-film](https://fontawesome.com/v4.7.0/icon/film).


**Reviewers**
- [x] @viadanna
- [x] @bradenmacdonald 
- [ ] edX reviewer[s] TBD

**Settings**
```yaml
EDXAPP_EXTRA_REQUIREMENTS:
  - name: git+https://github.com/open-craft/custom-unit-icons.git@agrendalath/bb-933-address_override_changes#egg=custom-unit-icons
  - name: git+https://github.com/edx/edx-django-utils.git@master#egg=edx-django-utils

OVERRIDE_GET_UNIT_ICON: custom_unit_icons.icons.get_icon
XBLOCK_EXTRA_MIXINS:
    - custom_unit_icons.icons.IconOverrideMixin
```

**TODO**:
- [ ] Provision the sandbox.
- [x] Update the docs from #24057.

edx-webhook commented:
Thanks for the pull request, @Agrendalath! I've created [OSPR-3801](https://openedx.atlassian.net/browse/OSPR-3801) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

natabene commented:
@Agrendalath Thank you for your contribution. Please let me know once it is ready for our review.

bradenmacdonald commented on openedx/core/lib/api/utils.py:
```diff
@@ -0,0 +1,41 @@
+            module, function = function_string.rsplit('.', 1)
+            mod = import_module(module)
+            func = getattr(mod, function)
+            return func(*args, **kwargs)
```
Maybe you should pass `f` into `func` so that `func` can optionally call the base implementation?

Also, it would be good if `DO_SOMETHING_IMPL` can be a list of functions, so that they can be chained. Otherwise, for example, as the platform grows, one might have two plugins installed that both want to add a custom XBlock icon handler. You could do something like:

```
prev_fn = f  # the base method in edx-platform
for impl in override_functions:
    prev_fn = partial(impl, prev_fn)
prev_fn()  # Call the last function in DO_SOMETHING_IMPL, and give it the chance to call the previous one, which can call the previous one, ..., which can call the base implementation. Or any one can return without calling the previous/base implementation.
```

Agrendalath replied:
That's a great idea!
Could you please check if [this](https://github.com/edx/edx-platform/pull/21433/commits/d67935154f8057d5737b9164de098d3d5074b441#diff-03c7b83e1f4c1cacc306a7966eaaf27cR47-R55) matches your idea, when you have a moment?

bradenmacdonald replied:
Looks great, thanks @Agrendalath !

bradenmacdonald commented on common/lib/xmodule/xmodule/x_module.py:
```diff
@@ -603,6 +605,14 @@ def get_child_by(self, selector):
+        An API for returning a css class identifying this module in the context of an icon.
+        It can be overridden by setting `GET_UNIT_ICON_IMPL` to an alternative implementation.
+        """
+        return self.get_icon_class()
```
If this is defined on `XModuleMixin`, it will only apply to `XModule` instances, not `XBlock` ones, which is the newer class. Since this is where `get_icon_class` is defined too, that's probably fine, but I don't think this needs to be an `XModuleMixin` method at all... I think it could just be a top-level method in some runtime/xblock utils file, e.g.

```python
@pluggable_override('GET_UNIT_ICON_IMPL')
def get_icon_for_block(block):
    return block.get_icon_class()
```

No need to change now, I'm not an official reviewer here, just pointing out options.

Agrendalath replied:
I was wondering about the correct place for this, but wasn't sure how this could be applied to `XBlock`, as it doesn't have a method for getting an icon. Should we just raise `NotImplementedError` when we don't find `block.get_icon_class` method?

bradenmacdonald replied:
I think it should be a top level function in [`xblock_utils/__init__.py`](https://github.com/edx/edx-platform/blob/master/openedx/core/lib/xblock_utils/__init__.py).

Agrendalath replied:
Sounds good, I'll move this.

bradenmacdonald commented on common/lib/xmodule/xmodule/modulestore/inheritance.py:
```diff
@@ -231,6 +231,13 @@ class InheritanceMixin(XBlockMixin):
         scope=Scope.settings
     )
 
+    icon = String(
```
`InheritanceMixin` is already kind of messy and we're trying not to make that worse. If this new `icon` setting is only for Units, then you should add this field to the `vertical` block directly, to avoid adding it to every other XBlock type when it's not needed.

Agrendalath replied:
Sure thing, I moved this to the `vertical` block.

Agrendalath commented:
@bradenmacdonald, thank you for your review! And sorry for addressing it late, I missed this email.

Agrendalath commented:
@natabene, this is ready for the review.

natabene commented:
@marcotuts Is this something you want to review before it moves to engineering?

bradenmacdonald commented on openedx/core/lib/api/utils.py:
```diff
@@ -19,23 +19,39 @@ def pluggable_override(override):
-            mod = import_module(module)
-            func = getattr(mod, function)
-            return func(*args, **kwargs)
+            prev_fn = f  # The base method in `edx-platform`.
```
This might need to be `partial(f, *args, **kwargs)` because it becomes the `prev_fn` arg later, and the other prev_fn args already have the parameters applied?

Agrendalath replied:
That's a good point. Thank you for catching this!

Agrendalath commented:
@marcotuts, @natabene, any news on this?

natabene commented:
@Agrendalath Sorry, no news yet.

edx-webhook commented:
@Agrendalath Even though your pull request wasn’t merged, please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/21433&entry.1484655318=https://github.com/Agrendalath&entry.752974735=2019-08-23+12:36&entry.1917517419=2020-02-12+07:51&entry.2133058324=No) so we can improve your experience in the future.

marcotuts commented:
@ormsbee - This is ok to move to engineering review. One question I have is that this seems like a large change to edx-platform for this plugin, is this largely a file relocation change, or what makes the change seem (at a glance) as large out of curiosity? Thanks!

Agrendalath commented:
@marcotuts, we divided monolithic `cms/static/js/views/modals/course_outline_modals.js` into smaller chunks to be able to override specific parts with a theme. This modular approach makes the theme much easier to maintain.

Btw. does the new courseware experience include these icons or is the navigation part altered in some way?

bradenmacdonald commented:
@marcotuts Does the new courseware experience include these icons?

<img width="871" alt="Screen Shot 2020-02-29 at 4 14 12 PM" src="https://user-images.githubusercontent.com/945577/75617034-8b1fac80-5b0e-11ea-980f-b6409b483bfa.png">

marcotuts commented:
@bradenmacdonald, @Agrendalath - The icons are used in the new learning sequence bar / MFE: https://github.com/edx/frontend-app-learning

bradenmacdonald commented:
@natabene  Per @marcotuts's comment above (thanks!), could you please update this to "Engineering Review"? I think the status got lost when we mistakenly closed and re-opened the PR.

> This is ok to move to engineering review.

natabene commented:
@bradenmacdonald The status is currently set to "Awaiting prioritization" and will be changed to "Engineering review" once we start reviewing it.

bradenmacdonald commented:
@Agrendalath Since the current courseware code (that you've modified here) will eventually be replaced by https://github.com/edx/frontend-app-learning , I think it might make sense to rebase this PR and strip out all the JavaScript/frontend changes, then test it with the frontend-app-learning MFE. That will make it easier to review and forwards-compatible.

Also I think it's worth calling more attention to the fact that this introduces a generic decorator for overriding any python function and as such provides a new model for platform extensibility - @natabene could we get some light engineering feedback on that approach soon?

Agrendalath commented:
@bradenmacdonald, sure, I'll set aside some time for the next sprint to change this.

natabene commented:
@nasthagiri Can we get early engineering feedback on approach soon for this?

bradenmacdonald commented:
@Agrendalath I'm working to document all of the platform's extension points in https://github.com/edx/edx-platform/pull/24057 . If that PR merges before this one, can you please update this one to add the `@pluggable_override` concept and the course icons override point to the new `extension_points.rst` documentation? And if yours merges first, I'll do that :)

Agrendalath commented:
Sure, @bradenmacdonald. Sounds good to me :)

Agrendalath commented:
@bradenmacdonald, I removed the frontend changes from this PR and opened  edx/frontend-app-learning#77 with the related LMS changes. We still need to find a way to plug the editor into new Studio MFE, but this should be clearer soon :)

@natabene, this is ready for the engineering feedback. The diff is much smaller now.

natabene commented:
@Agrendalath Thanks for letting me know, I will line this up for a review.

Agrendalath commented:
@bradenmacdonald, I added back an override in `seq_module.html`, because this feature would not work without a custom theme then. I also updated the testing instructions to include a way of manually updating an icon via modulestore, and added a TODO there.
I still need to find the proper sandbox, but some initial feedback (e.g. about the XBlocks vs XModules handling) would be appreciated.

bradenmacdonald commented on common/lib/xmodule/xmodule/vertical_block.py:
```diff
@@ -45,6 +50,13 @@ class VerticalBlock(SequenceFields, XModuleFields, StudioEditableBlock, XmlParse
+        default='default',
+        help=_("XBlock Icon"),
+        scope=Scope.settings,
+    )
```
We should avoid adding a new XBlock field here for all instances when most won't need this feature.

Instead, for any instances that want to allow overriding the icons of individual XBlocks (like this), we should just create a simple mixin like:

```python
class IconOverrideMixin:
    """ Allow changing the icon used for this specific instance of this XBlock """
    icon_override = String(
        display_name=_("Icon Override"),
        default='default',
        help=_("XBlock Icon Override"),
        scope=Scope.settings,
    )

XBLOCK_EXTRA_MIXINS = ["my_organization.customizations.IconOverrideMixin"]

XBLOCK_MIXINS = (LmsBlockMixin, InheritanceMixin, XModuleMixin, EditInfoMixin) + tuple(XBLOCK_EXTRA_MIXINS)
```

and add it to the instance configuration via the existing `XBLOCK_MIXINS` setting (I think it will need to be modified to allow specifying additional mixins, e.g. as `XBLOCK_EXTRA_MIXINS`).

Agrendalath replied:
That looks neat! Btw. is there a proper way to define these settings directly in a plugin, so we could just install it and skip changing these settings?

bradenmacdonald replied:
> Btw. is there a proper way to define these settings directly in a plugin, so we could just install it and skip changing these settings?

We could do that, yeah, though then there'd be no way to specify the order of things - which could be important for these type of chainable overrides. Overriding settings is also much easier these days now that [we don't have to modify `production.py` and `edx/configuration` every time we want to add/change a setting](https://github.com/edx/edx-platform/pull/24991#discussion_r490400696)

Agrendalath replied:
Yeah, it took me a moment to realize that I don't need to manually add new overrides from `ENV_TOKENS`, but I didn't dig how it's implemented. Thanks for pointing to this specific part of code, that's definitely good to know.

bradenmacdonald commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -519,6 +519,7 @@ def _render_student_view_for_items(self, context, display_items, fragment, view=
         for item in display_items:
             # NOTE (CCB): This seems like a hack, but I don't see a better method of determining the type/category.
             item_type = item.get_icon_class()
+            item_icon = item.get_icon()
```
Is `item_type` still used? It doesn't seem right that we call both `item.get_icon_class()` and `item.get_icon()` here, and use both results. I expected that `get_icon` would replace `get_icon_class` here, especially since by default `get_icon` just calls `get_icon_class`.

If we do need to keep both, perhaps rename the first one to `get_css_class()` ?

Agrendalath replied:
Yep, good point. I kept this for screen readers [here](https://github.com/edx/edx-platform/blob/9f39ca689d8ae5429a9ea77e34db1fde3e1ede44/lms/templates/seq_module.html#L74), but it makes more sense to override this everywhere. I remove the `item_icon` and replaced `get_icon_class` with `get_icon`.

bradenmacdonald commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -1262,6 +1262,10 @@ def create_xblock_info(xblock, data=None, metadata=None, include_ancestor_info=F
+        elif xblock.category == 'vertical':
+            xblock_info.update({
+                'icon': xblock.icon,
+            })
```
Why is this needed? It seems like there was no icon info here previously.

Agrendalath replied:
This was needed for Studio (additionally to the theme changes), but as we're extracting this field to a plugin, it sounds fine to remove it now and find an approach suitable for https://github.com/edx/frontend-app-course-authoring.

bradenmacdonald reviewed (commented):
Some comments. BTW we should remove a bunch of the boilerplate from https://github.com/open-craft/custom-unit-icons - for example, it definitely shouldn't specify any requirements unless they're in addition to the requirements already installed by openedx. It doesn't even really need to be a django plugin, although it doesn't hurt that it is.

Agrendalath commented:
@bradenmacdonald, I pushed the requested changes and opened a draft PR (https://github.com/open-craft/custom-unit-icons/pull/2) to `custom-unit-icons`. This requires some minor changes, but works.

bradenmacdonald reviewed (approved):
@Agrendalath Thanks for your patience and persistence here, and for the great work! I'm really excited about this new extension point, which should give flexibility for lots of little customizations without cluttering the codebase. (In fact, we already have a second use case, which lets us turn #21147 into a one-line PR to add this decorator, so the functionality can be kept as a plugin).

:+1: Approving but please incorporate the minor comments I've left here, then squash this down to a single commit with a nice commit message.

- [x] I tested this: as described (though I had to change the `XBLOCK_MIXINS` setting before step 2 would have any effect)
- [x] I read through the code
- [x] I checked for accessibility issues: n/a
- [x] Includes documentation: yes, thanks for updating `extension_points.rst`


@nasthagiri : do you or anyone else at edX want to take a look at this simple but powerful new extension point trial before we merge?

Agrendalath commented on docs/guides/extension_points.rst:
```diff
@@ -130,6 +130,9 @@ Here are the different integration points that python plugins can use:
      - Assess, Limited
      - XBlocks can also install test code that will then be run alongside the platform's usual python unit tests. It's unclear how well-supported this is at the moment.
+   * - Pluggable override (``openedx.core.lib.api.utils.pluggable_override``)
+     - Adopt, Stable
```
Ah, right, I was going to ask you about this one, but forgot to do this. Thanks for catching it :)

Agrendalath commented:
@bradenmacdonald, thank you for a very nice review and for tons of ideas to simplify and push forward this new way of extending edx-platform. That's great to hear that we already have more PRs where we can use this :)

I addressed your comments, then squashed and reworded commits.

> (though I had to change the XBLOCK_MIXINS setting before step 2 would have any effect)

Ah, good point. I missed reordering these.

nasthagiri commented:
@bradenmacdonald It would be great if @nedbat reviews this new capability: using Python decorators to redirect method calls.

To make this proposal even stronger, I suggest having a recommendation for namespacing these new Global constants.

nasthagiri commented on openedx/core/lib/api/utils.py:
```diff
@@ -0,0 +1,57 @@
+from django.conf import settings
+
+
+def pluggable_override(override):
```
One more comment:
It would be great if this general-purpose plugin functionality was co-located with our other general-purpose Plugin framework: https://github.com/edx/edx-django-utils/tree/master/edx_django_utils/plugins

Thanks!

Agrendalath replied:
Thanks for pointing to this, @nasthagiri! Should it be in the `utils.py` module there or in a separate one?

> To make this proposal even stronger, I suggest having a recommendation for namespacing these new Global constants.

Good point, I'll add this :+1:

nedbat commented on openedx/core/lib/api/utils.py:
```diff
@@ -0,0 +1,57 @@
+    This decorator allows overriding any function or method by pointing to an alternative implementation
+    with `override` param.
+    :param override: path to the alternative function
+
```
We should start out right, and give explicit instructions about how best to document uses of this decorator.  These will be introducing new settings that need explanataion.

Agrendalath replied:
Could you please give some more details about what you meant here?

Agrendalath replied:
@nedbat, I moved this change to edx/edx-django-utils#64, as @nasthagiri suggested. We can continue this conversation there.

idegtiarov commented on docs/guides/extension_points.rst:
```diff
@@ -142,6 +145,8 @@ Here are the different integration points that python plugins can use:
 .. |UserPartition docstring| replace:: ``UserPartition`` docstring
 .. _UserPartition docstring: https://github.com/edx/edx-platform/blob/f8cc58618a39c9f7b8e9e1001eb2d7a10395797e/common/lib/xmodule/xmodule/partitions/partitions.py#L105-L120
+.. |pluggable_override docstring| replace:: ``pluggable_override`` docstring
+.. _pluggable_override docstring: https://github.com/edx/edx-platform/blob/master/openedx/core/lib/api/utils.py
```
The link seems to be incorrect.

Agrendalath replied:
Correct, this is pointing to the `edx:master` branch, so it will be valid once we merge this.

idegtiarov replied:
Oh yes! Thanks for clarification. ))

Agrendalath commented on openedx/core/lib/api/utils.py:
```diff
@@ -0,0 +1,57 @@
+    1. Add this decorator to your original function. `DO_SOMETHING_IMPL` is the variable name in settings that can be
+       used for overriding this function.
+        >>> @pluggable_override('DO_SOMETHING_IMPL')
+        ...     def do_something(self):
```
@nedbat, indeed it is, thanks for catching it. I'll update it while addressing other comments.

> Actually, why are these REPL examples? That's not where people will be using this decorator.

I haven't found a better way to include some example code with syntax highlighting directly in the docstring. It might be possible to use the reStructured Text format here, but the syntax highlighting didn't work in my text editors then. I'm open to ideas here.

Agrendalath commented on openedx/core/lib/api/utils.py:
```diff
@@ -0,0 +1,57 @@
+        ...         return _something()
+
+    2. Prepare an alternative implementation:
+        >>> def do_something_different(prev_fn, something):
```
Sure, I'll make this more verbose.

natabene commented:
@Agrendalath @nedbat @bradenmacdonald Could you please clarify the status of this? Has the author implemented requested changes? Is review blocked by anything?

Agrendalath commented:
@natabene, thank you for checking. This is still blocked by the review of https://github.com/edx/edx-django-utils/pull/64. Would anybody from edX have the capacity to pick this up?
cc: @bradenmacdonald, @nedbat

natabene commented:
@Agrendalath Thanks for an update. I will check.

Agrendalath commented on requirements/edx/testing.txt:
```diff
@@ -105,7 +105,7 @@ edx-celeryutils==1.0.0    # via -r requirements/edx/base.txt, super-csv
 edx-django-release-util==1.0.0  # via -r requirements/edx/base.txt
 edx-django-sites-extensions==3.0.0  # via -r requirements/edx/base.txt
-edx-django-utils==3.15.0  # via -r requirements/edx/base.txt, django-config-models, edx-drf-extensions, edx-enterprise, edx-rest-api-client, edx-toggles, edx-when, ora2, super-csv
+-e git+https://github.com/open-craft/edx-django-utils.git@master#egg=edx-django-utils  # TODO: Change this to a proper version.
```
Once, edx/edx-django-utils#109 is merged, we will:
1. Revert this change.
1. Bump the required version from `edx-django-utils>=3.12.0` to `3.16.0` in `edx/base.in`.
1. Run `make upgrade` to pin the new version.

Agrendalath commented on requirements/edx/testing.txt:
```diff
@@ -105,7 +105,7 @@ edx-celeryutils==1.0.0    # via -r requirements/edx/base.txt, super-csv
 edx-django-release-util==1.0.0  # via -r requirements/edx/base.txt
 edx-django-sites-extensions==3.0.0  # via -r requirements/edx/base.txt
-edx-django-utils==3.15.0  # via -r requirements/edx/base.txt, django-config-models, edx-drf-extensions, edx-enterprise, edx-rest-api-client, edx-toggles, edx-when, ora2, super-csv
+-e git+https://github.com/edx/edx-django-utils.git@master#egg=edx-django-utils  # TODO: Change this to a proper version.
```
Once, edx/edx-django-utils#109 is merged, we will:
1. Revert this change.
1. Bump the required version from `edx-django-utils>=3.12.0` to `3.16.0` in `edx/base.in`.
1. Run `make upgrade` to pin the new version.

natabene commented:
@Agrendalath Is this good to review now? edx/edx-django-utils#64 has been merged

Agrendalath commented:
@natabene, yes, this is ready for review. We still need to push the new version of edx/edx-django-utils to PyPi (cf. my last comment on edx/edx-django-utils#109), but this is not a blocker for reviewing this.

natabene commented:
@bradenmacdonald Do you still want to review this as a Core Committer?

bradenmacdonald commented:
@natabene Yes, I'm hoping to get to it soon. Thanks!

bradenmacdonald commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -709,7 +712,7 @@ def _render_student_view_for_items(self, context, display_items, fragment, view=
         contents = []
         for item in display_items:
             # NOTE (CCB): This seems like a hack, but I don't see a better method of determining the type/category.
```
^ Is this old comment still needed?

Agrendalath replied:
@bradenmacdonald, probably not. I can remove it while squashing the commits.

bradenmacdonald reviewed (approved):
:+1: Looks great, thanks!

This is a conditional approval; before we merge, please:
* Get the version bump of `edx-django-utils` completed, and rebase this
* Squash down to a single commit following [conventional commits](https://github.com/edx/open-edx-proposals/blob/master/oeps/oep-0051-bp-conventional-commits.rst)

- [x] I tested this: on my devstack, as described
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

bradenmacdonald commented:
>  following [conventional commits](https://github.com/edx/open-edx-proposals/blob/master/oeps/oep-0051-bp-conventional-commits.rst)

Don't forget this part ;)

edx-webhook commented:
@Agrendalath 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/21433&entry.1484655318=https://github.com/Agrendalath&entry.752974735=2019-08-23+12:36&entry.1917517419=2021-04-07+13:42&entry.2133058324=Yes) so we can improve your experience in the future.

kdmccormick commented:
Thanks Braden and @Agrendalath . FYI, when squashing, I hard-wrapped your commit message body at ~80 chars. Git doesn't auto-wrap messages very well :)
